pkg/ccl/changefeedccl/kvfeed/kv_feed.go

```diff
@@ -472,6 +478,26 @@ func filterCheckpointSpans(spans []roachpb.Span, completed []roachpb.Span) []roa
 	return sg.Slice()
 }
 
+// filterCheckpointSpansFromCheckpoint filters spans which have already been
+// completed based on the given checkpoints information and returns the list of
+// spans that still need to be done. If checkpoint is nil, it will filter out
+// the given oldCheckpoint directly. Otherwise, it extracts and flattens spans
+// from newCheckpoint.Entries into a single slice and filters out the the
+// completed spans.
+func filterCheckpointSpansFromCheckpoint(
+	spansToScan []roachpb.Span, oldCheckpoint []roachpb.Span, newCheckpoint *jobspb.TimestampSpansMap,
+) []roachpb.Span {
+	if newCheckpoint == nil {
+		return filterCheckpointSpans(spansToScan, oldCheckpoint)
+	}
+
+	var spans []roachpb.Span
+	for _, sp := range newCheckpoint.Entries {
+		spans = append(spans, sp.Spans...)
```

Comment:
why is this impl not the same as filterCheckpointSpans, with a spangroup and sub? they contain the same info, no?

Reply:
Good idea, I refactored this part of the code to reuse `filterCheckpointSpans`. Let me know your thoughts. Unless I'm missing something this impl is slightly different than filterCheckpointSpans because `TimestampSpansMap` is a map of (timestamp -> []Span). We need to extract the spans and flatten into a single slice. Does that match your understanding?